Plugin causes critical issues profile swapping

tl;dr: on profile switch, task tracker can throw AssertionError, which is not caught by eventbus try/catch, so profile switching is incomplete, which can have problematic ramifications

if statement instead of assert here: https://github.com/osrs-reldo/tasks-tracker-plugin/blob/7a30111ad4950a37b57517eabf6c1b28b82243b4/src/main/java/net/reldo/taskstracker/panel/subfilters/FilterButtonPanel.java#L170 should resolve this problem

Context: 
I basically always use a jar of my hub plugin so things properly get tested, and because of this I have two different outcomes for this error.
The profile I have is a duplicate import of my main profile, but with changes I made during leagues, because of this it does _not_ use task-tracker

Outcome 1 - Using my plugins jar:
When attempting to profile swap while using my plugins jar, I fortunately get a stacktrace and the profile swap will appear to have worked e.g. settings are swapped, the different infoboxes from TimeTrackingReminder are a good example, but plugins from the previous swap are not changed.

<details>
  <summary>outcome1 stacktrace</summary>

```
2023-12-10 01:16:50 EST [pool-2-thread-1] INFO  n.r.client.config.ConfigManager - Switching profile to: actual default (42353538282200)
2023-12-10 01:16:50 EST [pool-2-thread-1] ERROR n.r.c.util.RunnableExceptionLogger - Uncaught exception in runnable net.runelite.client.plugins.config.ProfilePanel$$Lambda$2642/0x0000000801686670@17fb26d6
java.lang.AssertionError: null
        at net.reldo.taskstracker.panel.subfilters.FilterButtonPanel.redraw(FilterButtonPanel.java:170)
        at net.reldo.taskstracker.panel.subfilters.SkillFilterPanel.redraw(SkillFilterPanel.java:102)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at net.reldo.taskstracker.panel.SubFilterPanel.redraw(SubFilterPanel.java:42)
        at net.reldo.taskstracker.panel.LoggedInPanel.redraw(LoggedInPanel.java:113)
        at net.reldo.taskstracker.panel.TasksTrackerPluginPanel.redraw(TasksTrackerPluginPanel.java:63)
        at net.reldo.taskstracker.TasksTrackerPlugin.onConfigChanged(TasksTrackerPlugin.java:178)
        at net.runelite.client.eventbus.EventBus$Subscriber.invoke(EventBus.java:70)
        at net.runelite.client.eventbus.EventBus.post(EventBus.java:223)
        at net.runelite.client.config.ConfigManager.switchProfile(ConfigManager.java:234)
        at net.runelite.client.plugins.config.ProfilePanel.lambda$switchToProfile$7(ProfilePanel.java:758)
        at net.runelite.client.util.RunnableExceptionLogger.run(RunnableExceptionLogger.java:41)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

</details>

Outcome 2- Using the client as a normal user would:

When using the client normally, the issue presents more confusing for the user, but is handled more gracefully.

Swapping profiles will appear to do nothing as previously, but this time reloading the config panel will display as if the client has no profiles, waiting 10-15 seconds after this returns a full correct profile swap, no error is written to the logs.

<details>
  <summary>outcome2 demonstration</summary>

![RuneLite_2023-12-10_01-37-03](https://github.com/osrs-reldo/tasks-tracker-plugin/assets/41973452/48ef34fb-443e-44dd-82f3-364d35e8a704)

</details>

--- a/src/main/java/net/reldo/taskstracker/TasksTrackerConfig.java
+++ b/src/main/java/net/reldo/taskstracker/TasksTrackerConfig.java
@@ -91,12 +91,12 @@ public interface TasksTrackerConfig extends Config
 
 	@ConfigItem(
 		position = 106,
-		keyName = "taskTypeName",
+		keyName = "taskTypeJsonName",
 		name = "Task Type",
 		description = "Configures the task type which is displayed in the panel.",
 		hidden = true
 	)
-	default String taskTypeName()
+	default String taskTypeJsonName()
 	{
 		return "COMBAT";
 	}

--- a/src/main/java/net/reldo/taskstracker/TasksTrackerPlugin.java
+++ b/src/main/java/net/reldo/taskstracker/TasksTrackerPlugin.java
@@ -51,6 +51,7 @@ import net.runelite.client.config.ConfigManager;
 import net.runelite.client.config.RuneScapeProfileType;
 import net.runelite.client.eventbus.Subscribe;
 import net.runelite.client.events.ConfigChanged;
+import net.runelite.client.events.ProfileChanged;
 import net.runelite.client.game.SpriteManager;
 import net.runelite.client.plugins.Plugin;
 import net.runelite.client.plugins.PluginDescriptor;
@@ -118,8 +119,8 @@ public class TasksTrackerPlugin extends Plugin
 	{
 		try
 		{
-			String taskTypeName = config.taskTypeName();
-			taskService.setTaskType(taskTypeName);
+			String taskTypeJsonName = config.taskTypeJsonName();
+			taskService.setTaskType(taskTypeJsonName);
 		}
 		catch (Exception ex)
 		{
@@ -252,9 +253,15 @@ public class TasksTrackerPlugin extends Plugin
 		}
 	}
 
+	@Subscribe
+	public void onProfileChanged(ProfileChanged profileChanged)
+	{
+		reload();
+	}
+
 	public void refresh()
 	{
-		SwingUtilities.invokeLater(() -> this.pluginPanel.refresh(null));
+		SwingUtilities.invokeLater(() -> pluginPanel.refresh(null));
 	}
 
 	public void reloadTaskType()
@@ -263,8 +270,8 @@ public class TasksTrackerPlugin extends Plugin
 		filterService.clearFilterConfigs();
 		try
 		{
-			String taskTypeName = config.taskTypeName();
-			taskService.setTaskType(taskTypeName);
+			String taskTypeJsonName = config.taskTypeJsonName();
+			taskService.setTaskType(taskTypeJsonName);
 		}
 		catch (Exception ex)
 		{
@@ -272,8 +279,8 @@ public class TasksTrackerPlugin extends Plugin
 		}
 		SwingUtilities.invokeLater(() ->
 		{
-			this.pluginPanel.redraw();
-			this.pluginPanel.refresh(null);
+			pluginPanel.redraw();
+			pluginPanel.refresh(null);
 		});
 	}
 
@@ -312,7 +319,7 @@ public class TasksTrackerPlugin extends Plugin
 			return;
 		}
 
-		if (!reldoImport.taskTypeName.equalsIgnoreCase(config.taskTypeName()))
+		if (!reldoImport.taskTypeName.equalsIgnoreCase(config.taskTypeJsonName()))
 		{
 			this.showMessageBox("Import Tasks Error", String.format("Wrong task type. Select the %s task type to import this data.", reldoImport.taskTypeName), JOptionPane.ERROR_MESSAGE, false);
 			return;

--- a/src/main/java/net/reldo/taskstracker/TasksTrackerPlugin.java
+++ b/src/main/java/net/reldo/taskstracker/TasksTrackerPlugin.java
@@ -9,9 +9,11 @@ import java.awt.datatransfer.StringSelection;
 import java.awt.image.BufferedImage;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.EnumMap;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import javax.annotation.Nullable;
@@ -39,9 +41,12 @@ import net.reldo.taskstracker.panel.TaskTrackerPanelModule;
 import net.reldo.taskstracker.panel.TasksTrackerPluginPanel;
 import net.runelite.api.ChatMessageType;
 import net.runelite.api.Client;
+import net.runelite.api.Experience;
 import net.runelite.api.GameState;
+import net.runelite.api.Skill;
 import net.runelite.api.events.GameStateChanged;
 import net.runelite.api.events.GameTick;
+import net.runelite.api.events.StatChanged;
 import net.runelite.api.events.VarbitChanged;
 import net.runelite.client.callback.ClientThread;
 import net.runelite.client.chat.ChatMessageBuilder;
@@ -82,6 +87,7 @@ public class TasksTrackerPlugin extends Plugin
 	private long lastVarpUpdate = 0;
 	private NavigationButton navButton;
 	private RuneScapeProfileType currentProfileType;
+	private final Map<Skill, Integer> oldExperience = new EnumMap<>(Skill.class);
 
 	@Inject	@Named("runelite.version") private String runeliteVersion;
 	@Inject private Gson gson;
@@ -243,20 +249,48 @@ public class TasksTrackerPlugin extends Plugin
 			varpIdsToUpdate = new HashSet<>();
 			lastVarpUpdate = currentTimeEpoch;
 		}
+	}
 
+	@Subscribe
+	public void onStatChanged(StatChanged statChanged)
+	{
+		// @todo deprecate one of these, we don't need to track player skills twice.
+		// Cache current player skills
 		int[] newSkills = client.getRealSkillLevels();
 		boolean changed = !Arrays.equals(playerSkills, newSkills);
 		if (changed)
 		{
 			playerSkills = client.getRealSkillLevels();
-			SwingUtilities.invokeLater(() -> pluginPanel.refresh(null));
 		}
+
+		final Skill skill = statChanged.getSkill();
+
+		// Modified from m0bilebtw's modification from Nightfirecat's virtual level ups plugin
+		final int xpAfter = client.getSkillExperience(skill);
+		final int levelAfter = Experience.getLevelForXp(xpAfter);
+		final int xpBefore = oldExperience.getOrDefault(skill, -1);
+		final int levelBefore = xpBefore == -1 ? -1 : Experience.getLevelForXp(xpBefore);
+
+		oldExperience.put(skill, xpAfter);
+
+		// Do not proceed if any of the following are true:
+		//  * xpBefore == -1              (don't fire when first setting new known value)
+		//  * xpAfter <= xpBefore         (do not allow 200m -> 200m exp drops)
+		//  * levelBefore >= levelAfter   (stop if we're not actually reaching a new level)
+		//  * levelAfter > MAX_REAL_LEVEL (stop if above 99)
+		if (xpBefore == -1 || xpAfter <= xpBefore || levelBefore >= levelAfter || levelAfter > Experience.MAX_REAL_LEVEL)
+		{
+			return;
+		}
+
+		// If we get here, 'skill' was leveled up!
+		pluginPanel.taskListPanel.refreshTaskPanelsWithSkill(skill);
 	}
 
 	@Subscribe
 	public void onProfileChanged(ProfileChanged profileChanged)
 	{
-		reload();
+		reloadTaskType();
 	}
 
 	public void refresh()

--- a/src/main/java/net/reldo/taskstracker/data/task/TaskService.java
+++ b/src/main/java/net/reldo/taskstracker/data/task/TaskService.java
@@ -74,7 +74,7 @@ public class TaskService
 		{
 			tasks.clear();
 			currentTaskType = newTaskType;
-			configManager.setConfiguration(TasksTrackerPlugin.CONFIG_GROUP_NAME, "taskTypeName", newTaskType.getTaskJsonName());
+			configManager.setConfiguration(TasksTrackerPlugin.CONFIG_GROUP_NAME, "taskTypeJsonName", newTaskType.getTaskJsonName());
 
             // Complete creation of any GLOBAL value type filterConfigs
 			for (FilterConfig filterConfig : currentTaskType.getFilters())

--- a/src/main/java/net/reldo/taskstracker/panel/LoggedInPanel.java
+++ b/src/main/java/net/reldo/taskstracker/panel/LoggedInPanel.java
@@ -83,6 +83,22 @@ public class LoggedInPanel extends JPanel
 
 	public void redraw()
 	{
+		// taskTypeDropdown may become de-synced after profile change
+		String selectedTaskTypeJsonName = taskTypeDropdown.getItemAt(taskTypeDropdown.getSelectedIndex()).getValue().getTaskJsonName();
+		if(!selectedTaskTypeJsonName.equals(config.taskTypeJsonName()))
+		{
+			log.info("Task type dropdown de-synced, attempting to find current task type");
+			for(int i = 0; i < taskTypeDropdown.getItemCount(); i++)
+			{
+				ComboItem<TaskType> item = taskTypeDropdown.getItemAt(i);
+				if(item.getValue().getTaskJsonName().equals(config.taskTypeJsonName()))
+				{
+					log.info("Current task type found, setting selected task type");
+					taskTypeDropdown.setSelectedIndex(i);
+					break;
+				}
+			}
+		}
 		subFilterPanel.redraw();
 		sortPanel.redraw();
 		updateCollapseButtonText();

--- a/src/main/java/net/reldo/taskstracker/panel/TaskListPanel.java
+++ b/src/main/java/net/reldo/taskstracker/panel/TaskListPanel.java
@@ -50,32 +50,37 @@ public class TaskListPanel extends JScrollPane
 
 	public void refresh(TaskFromStruct task)
 	{
-		assert SwingUtilities.isEventDispatchThread();
-
-		emptyTasks.setVisible(false);
-
-		if (task != null)
+		if(SwingUtilities.isEventDispatchThread())
 		{
-			Optional<TaskPanel> panel = taskPanels.stream()
-				.filter(tp -> tp.task.getName().equalsIgnoreCase(task.getName()))
-				.findFirst();
-			panel.ifPresent(TaskPanel::refresh);
+			emptyTasks.setVisible(false);
+
+			if (task != null)
+			{
+				Optional<TaskPanel> panel = taskPanels.stream()
+					.filter(tp -> tp.task.getName().equalsIgnoreCase(task.getName()))
+					.findFirst();
+				panel.ifPresent(TaskPanel::refresh);
+			}
+			else
+			{
+				for (TaskPanel taskPanel : taskPanels)
+				{
+					taskPanel.refresh();
+				}
+			}
+
+			Optional<TaskPanel> visibleTaskPanel = taskPanels.stream()
+					.filter(TaskPanel::isVisible)
+					.findFirst();
+
+			if (visibleTaskPanel.isEmpty())
+			{
+				emptyTasks.setVisible(true);
+			}
 		}
 		else
 		{
-			for (TaskPanel taskPanel : taskPanels)
-			{
-				taskPanel.refresh();
-			}
-		}
-
-		Optional<TaskPanel> visibleTaskPanel = taskPanels.stream()
-				.filter(TaskPanel::isVisible)
-				.findFirst();
-
-		if (visibleTaskPanel.isEmpty())
-		{
-			emptyTasks.setVisible(true);
+			log.error("Task list panel refresh failed - not event dispatch thread.");
 		}
 	}
 
@@ -102,33 +107,39 @@ public class TaskListPanel extends JScrollPane
 		public void redraw()
 		{
 			log.debug("TaskListPanel.redraw");
-			assert SwingUtilities.isEventDispatchThread();
-			removeAll();
-			taskPanels.clear();
-			add(emptyTasks);
-			emptyTasks.setVisible(false);
-
-			log.debug("TaskListPanel creating panels");
-			List<TaskFromStruct> tasks = taskService.getTasks();
-			if (tasks == null || tasks.isEmpty())
+			if(SwingUtilities.isEventDispatchThread())
 			{
-				emptyTasks.setVisible(true);
-				return;
-			}
+				removeAll();
+				taskPanels.clear();
+				add(emptyTasks);
+				emptyTasks.setVisible(false);
 
-			for (int indexPosition = 0; indexPosition < tasks.size(); indexPosition++)
+				log.debug("TaskListPanel creating panels");
+				List<TaskFromStruct> tasks = taskService.getTasks();
+				if (tasks == null || tasks.isEmpty())
+				{
+					emptyTasks.setVisible(true);
+					return;
+				}
+
+				for (int indexPosition = 0; indexPosition < tasks.size(); indexPosition++)
+				{
+					int adjustedIndexPosition = indexPosition;
+					if (plugin.getConfig().sortDirection().equals(ConfigValues.SortDirections.DESCENDING))
+						adjustedIndexPosition = tasks.size() - (adjustedIndexPosition + 1);
+					TaskPanel taskPanel = taskPanelFactory.create(tasks.get(taskService.getSortedTaskIndex(plugin.getConfig().sortCriteria(), adjustedIndexPosition)));
+					add(taskPanel);
+					taskPanels.add(taskPanel);
+				}
+
+				log.debug("TaskListPanel validate and repaint");
+				validate();
+				repaint();
+			}
+			else
 			{
-				int adjustedIndexPosition = indexPosition;
-				if (plugin.getConfig().sortDirection().equals(ConfigValues.SortDirections.DESCENDING))
-					adjustedIndexPosition = tasks.size() - (adjustedIndexPosition + 1);
-				TaskPanel taskPanel = taskPanelFactory.create(tasks.get(taskService.getSortedTaskIndex(plugin.getConfig().sortCriteria(), adjustedIndexPosition)));
-				add(taskPanel);
-				taskPanels.add(taskPanel);
+				log.error("Task list panel redraw failed - not event dispatch thread.");
 			}
-
-			log.debug("TaskListPanel validate and repaint");
-			validate();
-			repaint();
 		}
 	}
 }

--- a/src/main/java/net/reldo/taskstracker/panel/TaskListPanel.java
+++ b/src/main/java/net/reldo/taskstracker/panel/TaskListPanel.java
@@ -13,9 +13,11 @@ import javax.swing.border.EmptyBorder;
 import lombok.extern.slf4j.Slf4j;
 import net.reldo.taskstracker.TasksTrackerPlugin;
 import net.reldo.taskstracker.config.ConfigValues;
+import net.reldo.taskstracker.data.jsondatastore.types.TaskDefinitionSkill;
 import net.reldo.taskstracker.data.task.TaskFromStruct;
 import net.reldo.taskstracker.data.task.TaskService;
 import net.reldo.taskstracker.panel.components.FixedWidthPanel;
+import net.runelite.api.Skill;
 import net.runelite.client.ui.FontManager;
 
 @Slf4j
@@ -82,6 +84,29 @@ public class TaskListPanel extends JScrollPane
 		{
 			log.error("Task list panel refresh failed - not event dispatch thread.");
 		}
+	}
+
+	public void refreshTaskPanelsWithSkill(Skill skill)
+	{
+		// Refresh all task panels for tasks with 'skill' or
+		// 'SKILLS' (any skill) or 'TOTAL LEVEL' as a requirement.
+		taskPanels.stream()
+				.filter(tp ->
+				{
+					List<TaskDefinitionSkill> skillsList = tp.task.getTaskDefinition().getSkills();
+					if(skillsList == null || skillsList.isEmpty())
+					{
+						return false;
+					}
+
+					return skillsList.stream()
+						.map(TaskDefinitionSkill::getSkill)
+						.anyMatch(s ->  s.equalsIgnoreCase(skill.getName()) ||
+                                        s.equalsIgnoreCase("SKILLS") ||
+                                        s.equalsIgnoreCase("TOTAL LEVEL")
+						);
+				})
+				.forEach(TaskPanel::refresh);
 	}
 
 	private class TaskListListPanel extends FixedWidthPanel

--- a/src/main/java/net/reldo/taskstracker/panel/TasksTrackerPluginPanel.java
+++ b/src/main/java/net/reldo/taskstracker/panel/TasksTrackerPluginPanel.java
@@ -66,24 +66,29 @@ public class TasksTrackerPluginPanel extends PluginPanel
 
 	public void setLoggedIn(boolean loggedIn)
 	{
-		assert SwingUtilities.isEventDispatchThread();
-
-		if (loggedIn != this.loggedIn)
+		if(SwingUtilities.isEventDispatchThread())
 		{
-			if (loggedIn)
-			{
-				loggedOutPanel.setVisible(false);
-				loggedInPanel.setVisible(true);
-			}
-			else
-			{
-				loggedInPanel.setVisible(false);
-				loggedOutPanel.setVisible(true);
-			}
+            if (loggedIn != this.loggedIn)
+            {
+                if (loggedIn)
+                {
+                    loggedOutPanel.setVisible(false);
+                    loggedInPanel.setVisible(true);
+                }
+                else
+                {
+                    loggedInPanel.setVisible(false);
+                    loggedOutPanel.setVisible(true);
+                }
 
-			validate();
-			repaint();
+                validate();
+                repaint();
+            }
+            this.loggedIn = loggedIn;
 		}
-		this.loggedIn = loggedIn;
+		else
+		{
+			log.error("Failed to update loggedIn state - not event dispatch thread.");
+		}
 	}
 }

--- a/src/main/java/net/reldo/taskstracker/panel/filters/DynamicDropdownFilterPanel.java
+++ b/src/main/java/net/reldo/taskstracker/panel/filters/DynamicDropdownFilterPanel.java
@@ -70,18 +70,23 @@ public class DynamicDropdownFilterPanel extends FilterPanel
 
 	public void redraw()
 	{
-		assert SwingUtilities.isEventDispatchThread();
+		if(SwingUtilities.isEventDispatchThread())
+		{
+			removeAll();
 
-		removeAll();
+			dropdown = makeDropdownPanel();
 
-		dropdown = makeDropdownPanel();
+			add(makeLabel());
+			add(dropdown);
 
-		add(makeLabel());
-		add(dropdown);
+			updateFilterConfig();
 
-		updateFilterConfig();
-
-		validate();
-		repaint();
+			validate();
+			repaint();
+		}
+		else
+		{
+			log.error("Dropdown filter panel redraw failed - not event dispatch thread.");
+		}
 	}
 }

--- a/src/main/java/net/reldo/taskstracker/panel/filters/FilterButtonPanel.java
+++ b/src/main/java/net/reldo/taskstracker/panel/filters/FilterButtonPanel.java
@@ -19,12 +19,14 @@ import javax.swing.SwingUtilities;
 import javax.swing.border.EmptyBorder;
 import javax.swing.plaf.basic.BasicBorders;
 import javax.swing.plaf.basic.BasicButtonUI;
+import lombok.extern.slf4j.Slf4j;
 import net.reldo.taskstracker.TasksTrackerPlugin;
 import net.runelite.client.ui.ColorScheme;
 import net.runelite.client.ui.FontManager;
 import net.runelite.client.util.ImageUtil;
 import net.runelite.client.util.SwingUtil;
 
+@Slf4j
 public abstract class FilterButtonPanel extends FilterPanel
 {
     protected final TasksTrackerPlugin plugin;
@@ -194,23 +196,28 @@ public abstract class FilterButtonPanel extends FilterPanel
 
     public void redraw()
     {
-        assert SwingUtilities.isEventDispatchThread();
+        if(SwingUtilities.isEventDispatchThread())
+        {
+            buttons.clear();
+            removeAll();
 
-        buttons.clear();
-        removeAll();
+            collapseBtn = makeCollapseButton();
+            buttonPanel = makeButtonPanel();
 
-        collapseBtn = makeCollapseButton();
-        buttonPanel = makeButtonPanel();
+            add(collapseBtn, BorderLayout.NORTH);
+            add(buttonPanel, BorderLayout.CENTER);
+            add(allOrNoneButtons(), BorderLayout.SOUTH);
+            updateFilterText();
+            updateCollapseButtonText();
 
-        add(collapseBtn, BorderLayout.NORTH);
-        add(buttonPanel, BorderLayout.CENTER);
-        add(allOrNoneButtons(), BorderLayout.SOUTH);
-        updateFilterText();
-        updateCollapseButtonText();
+            collapseBtn.setVisible(plugin.getConfig().filterPanelCollapsible());
 
-        collapseBtn.setVisible(plugin.getConfig().filterPanelCollapsible());
-
-        validate();
-        repaint();
+            validate();
+            repaint();
+        }
+        else
+        {
+            log.error("Filter button panel redraw failed - not event dispatch thread.");
+        }
     }
 }

